[AI] [Micro AI] [1.17] Add new Micro AI: ai_type=bloodlust

knyghtmare opened the pull request:
Here is a subset of the hunter MAI, called `ai_type=boss`, intended for dungeon and scenario boss-category units.

Sample Code:
```ini
[micro_ai]
        side=2
        ai_type=boss
        action=add

        id="Mal Keshar"
        [filter_location]
            x=30
            y=31
            radius=6
        [/filter_location]
[/micro_ai]
```

What it does?

1. Makes the 'boss' wander around a specified area (boss arena)
2. If enemy is within the boss's movement, target weakest enemy.
3. Approach and attack the weakest enemy.
4. Kill and reset to wandering mode.

Pending Stuff before making a full PR:

- [x] Test map
- [x] Test Scenario
- [ ] Documentation when merged.
- [x] Update Code to accomodate LuaAPI changes

Closes #6372

stevecotton commented:
The name "boss" doesn't seem descriptive to me - I'd expect "boss" to be something like the standard AI's behavior for leaders, where if they're not on a castle they get on a castle, but if they are on a castle they might attack. I'd also suggest searching the forums and seeing if the name hits a lot of unrelated topics, currently "boss" finds 2152 existing posts.

Not sure if "gladiator" would be better, it's what the description in this PR suggests to me, but it's also a term that's been heavily used in the forums with 466 matching posts.

knyghtmare commented:
> Not sure if "gladiator" would be better, it's what the description in this PR suggests to me, but it's also a term that's been heavily used in the forums with 466 matching posts.

TBH, the current behaviour is a hyper aggressive unit (or group of them) who will ruthlessly attack enemy units within the stated area. Possible tentative names:

- scenario_boss
- gladiator
- berserker
- aggressor
- hunter_no_home

knyghtmare commented:
I have requested reviews from octalot (as the current AI project member is unfilled) and CelticMinstrel (as he's the expert on Lua)

CelticMinstrel commented on data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:
```diff
@@ -25,7 +25,44 @@
     {DEFAULT_MUSIC_PLAYLIST}
 
-    map_file=ai/micro_ais/maps/dungeon_mai_comparisons.map
+    # map_file=ai/micro_ais/maps/dungeon_mai_comparisons.map
```
I'm guessing this would work if you add `data/` in front.

knyghtmare replied:
I rectified this when I found it was due to me not pushing the changes from my fork to the PR and then pulling to my local build

CelticMinstrel commented:
I don't see any major problems with this. I haven't taken a close look at the actual AI code though.

CelticMinstrel commented:
Is the MAI abbreviation used in other test scenarios? I think it's better not to use it, but if others already do, then whatever.

knyghtmare commented:
> Is the MAI abbreviation used in other test scenarios? I think it's better not to use it, but if others already do, then whatever.

Yeah, good point. I have to change the scenario id and name to something less technical, in like 2 words or less.

stevecotton commented on data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:
```diff
@@ -0,0 +1,294 @@
+        [message]
+            speaker=narrator
+            image=wesnoth-icon.png
+            message= _ "The boss micro AI has been assigned to a boss-class unit along with 2 other identical units which have the Zone Guardian Micro AI and the Hunter MAI respectively. This was done so was to investigate any behavioural changes and similaries can be observed."
```
.. so as to ...

knyghtmare replied:
Fixed

stevecotton commented on data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:
```diff
@@ -0,0 +1,294 @@
+        [message]
+            speaker=narrator
+            image=wesnoth-icon.png
+            message= _ "The Boss Micro AI will initially wander around in its arena. However, if an enemy unit is within its zone, it will approach and target the unit. On kill, the boss unit will check for any nearby targets. If any are found, it will attack those. If not, he will return to wandering around in its zone. If an enemy unit decides to run away, the boss unit will pursue and will not ...
```
There's a pronoun mismatch in there ("he" as opposed to "it"). Trivial stuff, but might as well comment on it.

knyghtmare replied:
fixed.

stevecotton commented:
Looks reasonable to me, but the AI is an area that's unfamiliar to me.

Please add it to data/ai/micro_ais/scenarios/scenario_micro_ai.cfg.

knyghtmare commented:
> Looks reasonable to me, but the AI is an area that's unfamiliar to me.
> 
> Please add it to data/ai/micro_ais/scenarios/scenario_micro_ai.cfg.

Okay, I have made the adjustments. What else is left to do before squash merging (or, I do need squash merging or not?).

soliton- commented:
Commits should be squashed if they're not very useful history. Like typos introduced in one commit and fixed in a later one or pretty much any other instances of fixing up earlier commits. Or just plain bad commits like the "Changed file X" ones you get from website commits when you don't actually write a commit message. If none of those exist or were already squashed and force pushed there is no reason to further squash good commits.

CelticMinstrel commented:
As far as I can tell, there's no good reason for this change to have 23 commits (most of them are fixups, reverts, or minor addressing feedback type of commits), so I'd just use the squash and merge button, which also means that @knyghtmare doesn't need to do any work to squash stuff.

knyghtmare commented:
A squash and merge would be fine. Seems appropriate for this PR.

stevecotton commented on data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:
```diff
@@ -0,0 +1,295 @@
+        immutable=yes
+        side=0
+        text= _ "Zone Guardian MAI Chamber"
+       
```
One of the reasons that the CI is failing is that wmlindent needs to be run on this file (or the whitespace fixed).

stevecotton commented on data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:
```diff
@@ -0,0 +1,295 @@
+    [label]
+        color="164,161,221"
+        immutable=yes
+        side=0
```
The schema is also missing `side`. I think it's easiest to just delete it from this file for now.

CelticMinstrel replied:
I think 0 is also the default?

stevecotton commented:
> If enemy is within the defined area, target weakest enemy.

That doesn't seem to be what it does - the defined area only seems to affect the hexes that the boss may choose for random movement. The AI seems to ignore enemies within the defined area if the boss can't reach them in the current turn.

stevecotton commented on data/ai/micro_ais/maps/dungeon_mai_comparisons.map:
```diff
@@ -0,0 +1,35 @@
+Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Cud^Xo, Cud^Ebn, Cud^Xo, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu
+Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Cud^Xo, Cud^Xo, Cud^Xo, Cud^Xo, Cud^Xo, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu
+Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Cud^Xo, Cud^Xo, Urb, Ur, Ias, Uu, Uu, Cud^Xo, Cud^Xo, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu
+Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Cud^Xo, Cud^Xo, Cud^Xo, Cud^Xo, Uu, Urb, Uu, Urb, Ias, Ur, Ur, Urb, Urb, Cud^Xo, Cud^Xo, Cud^Xo, Cud^Xo, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu
```
The map is decorative but doesn't give a good illustration of the bosses' areas. I assumed that each courtyard would be the bosses' area, but the areas are (radius 5) which the courtyards are (radius 6 with 6 corners removed).

knyghtmare replied:
The map was made like that to exhibit the case of the elite-class units behavior on units which are in the zone and within or outside their movement range.

stevecotton replied:
I'm don't understand. The visual clues on the map are making me think the zones are different to their actual hexes.

What are the castle hexes around the zones meant to mean?

knyghtmare replied:
The castle hexes around the zones dont serve any role other than to further augment the appearance of an arena. Aesthetics.

stevecotton replied:
For the micro ai demo scenarios, I think understanding is more important than aesthetics. You could do with labels on the map, but please communicate the AI's zones to the person viewing the map.

knyghtmare commented:
> > If enemy is within the defined area, target weakest enemy.
> 
> That doesn't seem to be what it does - the defined area only seems to affect the hexes that the boss may choose for random movement. The AI seems to ignore enemies within the defined area if the boss can't reach them in the current turn.

Oops. I mistyped that one. My bad. Rectified. Same for the in-scenario note/description.

stevecotton commented:
The "will pursue and will not stop" sounds like it fixes on a victim, like The Terminator or various horror monsters. The actual behavior is that it'll return to its zone if the victim manages to run outside its movement zone.

Suggestion: rename it something like "sulking bloodlust". If any enemy is within this units' attack range, it will attack the weakest, even if this requires it to move away from its assigned zone. If no enemies are within range, it will wander randomly within its assigned zone, returning there if necessary.

knyghtmare commented:
> Suggestion: rename it something like "sulking bloodlust". If any enemy is within this units' attack range, it will attack the weakest, even if this requires it to move away from its assigned zone. If no enemies are within range, it will wander randomly within its assigned zone, returning there if necessary.

That works. Also, is it implied to rename it from `ai_type=boss` to `ai_type=bloodlust`?
The name sounds appropriate

stevecotton commented on data/ai/micro_ais/cas/ca_bloodlust.lua:
```diff
@@ -0,0 +1,132 @@
+
+    -- If bloodlust_unit_status is not set for the bloodlust_unit -> default behavior -> random wander
+    if (not bloodlust_unit_vars.bloodlust_unit_status) then
+        -- bloodlust_unit gets a new goal if none exist or on any move with 10% random chance
```
A few of the comments have been hit by a find-and-replace, and should possible just say "unit" or similar.

stevecotton reviewed (approved):
Looks good. My only comment is a trivial issue, which you can decide to ignore and merge as-is.

knyghtmare commented:
I did a last check and I fixed two things.
1. an image was not showing as I assumed the Ruffian Unit was an outlaw unit and not a peasant (which it actually is).
2. I have added a note indicating that the testers have access to three custom right-click menus for probing the AI behaviour.

I think I should have added that note earlier so that the missing image thing would have been revealed sooner.
Initially I speculated that the image not showing was caused by the crop IPF but the image path itself was mis-typed.

knyghtmare commented:
Here is also my draft of the bloodlust MAI's wiki entry:
## Bloodlust MAI
The Bloodlust Micro AI leads a unit (or units) through behavior analogous to hyper aggressive or boss-class unit(s). The units wander in a prescribed area of the map, their "arena" zone, until an enemy unit comes within range. Note that these units do not actively chase down enemies when they enter the zone, but wander randomly through the arena zone until an enemy unit is encountered.
 If enemies are within range of the bloodlust-MAI unit, this unit attacks the weakest of them. If a kill is made, the unit resets its status and either roams the arena again (if no enemies are within range) or targets another of the weakest enemies if they are within attack range.

Note: this Micro AI is best optimized for dungeon bosses/minibosses where arena zone definitions are most prominent or for portraying the behavior of "berserker" units which are too frenzied to either see reason or cooperate with units belonging to the same side.

Is the draft good to go?

stevecotton commented on data/ai/micro_ais/cas/ca_bloodlust.lua:
```diff
@@ -0,0 +1,132 @@
+    local bloodlust_unit_vars = MAIUV.get_mai_unit_variables(bloodlust_unit, cfg.ai_id)
+
+    -- If bloodlust_unit_status is not set for the bloodlust_unit -> default behavior -> random wander
+    if (not bloodlust_unit_vars.bloodlust_unit_status) then
```
I've only just realised that `bloodlust_unit_status` is only set to `nil`. Unlike the Hunter CA's `hunting_status`, this variable doesn't seem to have a purpose.

It might be counting as a change for the purpose of checking whether the execution function has changed the game-state, but if it is then that would likely be a bug.

stevecotton reviewed (changes requested):
> If enemies are within range of the bloodlust-MAI unit, this unit attacks the weakest of them. If a kill is made, the unit resets its status and either roams the arena again (if no enemies are within range) or targets another of the weakest enemies if they are within attack range.

Sorry for flip-flopping on the approval, but this paragraph made me notice something about the code. This description suggests that there's some logic of locking-on to the choice of target enemy, but AFAICS the only logic in the code that affects future turns is during the wandering cycle, when it selects an x,y pair to wander to.

CelticMinstrel commented:
Since the AI was renamed, could you update the PR title?

stevecotton commented:
Currently I think this is waiting on @knyghtmare to respond to the review about `bloodlust_unit_status`. No rush, this comment is just to keep track of what is waiting on what.

Pentarctagon commented:
@knyghtmare any updates on this?

knyghtmare commented:
> any updates on this?

I believe it was later decided that the MAI would be not necessary and a `return_home=true/false` bool key would be added to the hunter MAI instead which would give the same behaviour.

In any case, these Bloodlust MAI and Minions MAI will be coming packaged in NR-rework. The decision to add to core or not can come later.

Pentarctagon commented:
Alright, so it sounds like this PR can be closed since it'll be superseded by the NR rework?

Pentarctagon commented:
*pokes @knyghtmare*

Pentarctagon commented:
Thanks.
